### DISCUSSION TITLE
Data spill to the next chunk unexpectedly

### DISCUSSION THREAD

**Discussion by michaelfung** [mojolicious/mojo#1904] [Q&A] [answered]
I use chunked transfer with code like the documentation. When transferring thousands of lines of JSON string e.g. `{"user_id":123456}`. I send 100 lines per chunk and each line is less than 20 bytes. 

The problem is: at the receiving end (use Mojo::UserAgent) randomly some lines failed to decode as they are split between chunks. So, is this behavior normal or a bug?

```
# Sending side:
my $drain = sub {
  my $c = shift;
  my $chunk = '';
  for ( 1 .. 100 ) {
      last unless my $row = $users[$i++];
      $chunk .= encode_json( $row ) . "\n";
  }
  return $c->finish if ($chunk eq '');
  $c->write_chunk( $chunk, __SUB__ );
};
$c->$drain;

# Receving side:
$tx->res->content->unsubscribe('read')->on(read => sub ($mojo_content_obj, $chunk) {
    my $chunk_row = 0;
    open my $fh, '<', \$chunk;
    while (my $line = <$fh>) {
      ++$chunk_row;
      my $row = eval { decode_json $line; };
      warn "bad row at beginning of chunk, line=$line\n" if $@ and $chunk_row == 1;
      ...
  }
```

**Comment by Grinnz** [comment]
If you do a chunked read you have to buffer the received bytes until you receive a full JSON object to decode

**Reply by michaelfung** [reply]
Each line is a complete JSON string. Sorry my post is not clear on that.

**Reply by Grinnz** [reply]
Line buffering chunked bytes is a common operation. See how it's done in https://metacpan.org/module/Mojo::IOLoop::Stream::Role::LineBuffer/source.

```perl
my $buffer = '';
$tx->res->content->unsubscribe('read')->on(read => sub ($mojo_content_obj, $chunk) {
  $buffer .= $chunk;
  while ($buffer =~ s/^(.*\n)//) {
    my $line = $1;
  }
});
```

**Reply by michaelfung** [reply]
From the doc of Mojo::Content, the **read** event:

**Emitted when a new chunk of content arrives.**

It leads me to expect a complete chunk. Perhaps the doc needs clarification?

**Reply by Grinnz** [reply]
What is a "complete chunk"?

**Reply by michaelfung** [reply]
100 lines of JSON strings, as in my case.

**Reply by Grinnz** [reply]
There is a max buffer size per write. You can't rely on receiving packets of a certain size.

**Reply by michaelfung** [reply]
I know. I will dig into Mojo::Content source code and see if it is possible to make it emit **read** only when a complete chunk is received. That will make memory requirements acceptable when transferring large number of lines.

**Reply by Grinnz** [reply]
Your concept of "complete chunk" no longer exists once it has left the source TCP socket.

**Reply by michaelfung** [reply]
According to the specification, there will be size in the head and a delimiter **\r\n** at the end of each chunk, I think we can make use of it to tell a complete chunk.

**Reply by Grinnz** [reply]
Ah, you are referring to chunked transfer encoding. Perhaps you are being mislead by the use of the word "chunk" in the read event description, but it operates on each chunk of bytes that arrives, unrelated to transfer encoding.

**Reply by michaelfung** [reply]
I would like to know what is the rationale for not doing emit "read" event only when whole chunk(s) is received?

It works for my use case after the following patch to Mojo::Content.

```
--- Content.pm.orig     2022-01-13 06:06:09.354216441 +0000
+++ Content.pm  2022-01-14 02:11:24.701287118 +0000
@@ -102,9 +102,11 @@
 
   # Chunked or relaxed content
   if ($self->is_chunked || $self->relaxed) {
-    $self->_decompress($self->{buffer} //= '');
+    if ($self->{whole_chunk_read}) {
+      $self->_decompress($self->{buffer} //= '') ;
+      $self->{buffer} = '';
+    }
     $self->{size} += length $self->{buffer};
-    $self->{buffer} = '';
     return $self;
   }
 
@@ -219,6 +221,7 @@
     $self->{buffer} .= substr $self->{pre_buffer}, 0, $len, '';
     $self->{real_size} += $len;
     $self->{chunk_len} -= $len;
+    $self->{whole_chunk_read} = ($self->{chunk_len} == 0) ? 1 : 0;
   }
```

**Reply by Grinnz** [reply]
Most HTTP content is not sent using chunked transfer encoding.

**Reply by michaelfung** [reply]
Hmm...??? Thanks anyway!

**Comment by michaelfung** [comment, marked as answer]
For the record:

> it operates on each chunk of bytes that arrives, unrelated to transfer encoding.
